drive_to_gps/src/calculate_twist_server.cpp

```diff
@@ -0,0 +1,92 @@
+#include "calculate_degrees.h"
+#include "drive_to_gps/dist_heading.h"
+#include "drive_to_gps/heading.h"
+#include "geometry_msgs/Twist.h"
+#include "ros/ros.h"
+#include "sensor_msgs/NavSatFix.h"
+#include <actionlib/server/simple_action_server.h>
+#include <drive_to_gps/gps_goalAction.h>
+#include <ros/callback_queue.h>
+
+#define MAX_ERROR 5         // error for distance comparisons in metres
+#define MAX_LINEAR_VEL 1.0  // max linear velocity of the rover (change)
+#define MAX_ANGULAR_VEL 1.0 // max angular velocity of the rover (change)
+
+typedef actionlib::SimpleActionServer<drive_to_gps::gps_goalAction> Server;
+
+ros::Publisher pub;
+ros::CallbackQueue my_queue;
+drive_to_gps::headingPtr curr_head = NULL; // current heading of the rover
+sensor_msgs::NavSatFixPtr curr_gps = NULL; // current gps coordinates
+int degrees_to_target = 0;
+
+void execute(const drive_to_gps::gps_goalGoalConstPtr &goal, Server *as) {
+  drive_to_gps::gps_goalFeedback feedback;
+  if (curr_gps == NULL || curr_head == NULL) {
+    ros::Rate r(1);
+    ROS_INFO("Waiting for topics to publish data");
+    while (ros::ok() && (curr_gps == NULL || curr_head == NULL)) {
+      my_queue.callAvailable();
+      r.sleep();
+    }
+  }
+  sensor_msgs::NavSatFixPtr gps_goal(
+      new sensor_msgs::NavSatFix(goal->gps_goal));
+  bool end = false;
+  ros::Rate rate(1);
+  while (calculate_distance(gps_goal, curr_gps) > MAX_ERROR && !end &&
+         ros::ok()) {
+    ROS_INFO("Calculating degrees, heading and sending twist message");
```

Comment:
A good practice is to always used the named variants of the rosconsole macros. the node can be written in a class and the constructor can take a name. this name can be used for all the rosconsole prints and the nodehandle. it keeps the code tidy and makes it easier during runtime to filter through all the topics and console prints